highfidelity/hifi
# Entity whitelist for raypicking and whiteboard painting example

This PR adds support for whitelisting entities to be raypicked. If the user provides an array of one or more entityIDs to as the 3rd argument of Entities.findRayIntersection() call, only entities contained in that array will be whitelisted and picked against. Otherwise they will be ignored.

This helps with raypicking performance when user knows there are certain items they want to pick against, and other entities in the scene can be ignored. For example, create a whiteboard in a complex scene such as toybox by runnning whiteboardSpawner.js. Now, in whiteboardEntityScript.js, uncomment line 71 and comment in line 73. The script will now pick against all entities in the scene, including complex models with many thousands of unsorted triangles, twice each frame (most likely at 60 fps- yikes!) as opposed to only picking the whiteboard. A noticeable increase in judder in the Rift becomes apparent.

The whiteboard scripts included in the PR comprise an app, that, when whiteboardSpawner is run, place a whiteboard in front of the client's avatar which ran the script. Any user can now approach the whiteboard and draw on it using their hydras. Users can switch colors by clicking or hydra grabbing the color boxes lining the top of the whiteboard. 

To test this script, run 
https://rawgit.com/ericrius1/hifi/whiteboard/examples/painting/whiteboard/whiteboardSpawner.js
 and walk up to the whiteboard and start painting with either hydra by holding z-trigger. (Make sure to test with this PR build so whitelist is honored)
